src/include/platform/GeneralFaults.h

```diff
@@ -0,0 +1,139 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          General faults could be detected by the Node.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPError.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+static constexpr size_t kMaxHardwareFaults = 11;
+static constexpr size_t kMaxRadioFaults    = 7;
+static constexpr size_t kMaxNetworkFaults  = 4;
+
+template <size_t N>
+class GeneralFaults
+{
+public:
+    /* The iterator */
+    class Iterator
+    {
+    public:
+        Iterator(const GeneralFaults<N> * GeneralFaults, int index);
+        uint8_t operator*() const;
+        Iterator & operator++();
+        bool operator!=(const Iterator & other) const;
+
+    private:
+        const GeneralFaults<N> * mGeneralFaultsPtr;
+        int mIndex = -1;
+    };
+
+public:
+    GeneralFaults() = default;
+    ~GeneralFaults() { mSize = 0; }
+
+    CHIP_ERROR add(const uint8_t value);
+
+    int size() const;
+    uint8_t operator[](int index) const;
+
+    Iterator begin() const;
+    Iterator end() const;
+
+private:
+    uint8_t mData[N];
+    int mSize = 0;
```

Comment:
Use `size_t`? Or `uint8_t` to reduce the memory usage.

Reply:
we need to use -1 to indicate invalid index